### PR TITLE
FileIO "bypass" support [ESD-1203]

### PR DESCRIPTION
**Pull request by silverjam**
## Design notes

This adds a new mode to `endpoint_adapter` and `sbp_fileio_daemon` that allow it to bypass the router, thus removing the need for the complications associated with trying to "fast pass" data through the router.  By adding server sockets to `sbp_fileio_daemon` we can also solve the problem of only responding to fileio data on the port it was received on, which avoids the issue where FileIO response data for a TCP connection would flood the UART.

## Testing

Mostly bench testing with HITL spot checks.  Seeing that this is actually the fastest of all previous methods because it allows FileIO response data to be batched into a large TCP send (rather than a bunch of small ones).  When data flowed through the endpoint_router, then to the `endpoint_adapter` we were forced through a framer at some point, which caused the response data to be cut up into smaller TCP packets.

## To-do
- [x] Clean-up metrics
- [x] Fix short option parsing for sbp_fileio_daemon
- [x] ~Decide if TCP client data transfer sockets are necessary~ -> turning off fileio "bypass" ports for tcp client connections, supporting it for uart, usb and tcp_server.
- [x] ~Test UART / USB data transfer~ -> Seems to work fine now

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019032221-17-g23ebcf52

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend runs](https://gnss-analysis.swiftnav.com/jobs/id=973678)
+ [live-roof-650-townsend-dropouts-zero-baseline runs](https://gnss-analysis.swiftnav.com/jobs/id=973679)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-17-g23ebcf52&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-17-g23ebcf52&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-17-g23ebcf52&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-17-g23ebcf52&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019032221-17-g23ebcf52
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019032221-17-g23ebcf52)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019032221-17-g23ebcf52/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019032221-17-g23ebcf52

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019032221-22-g062fb610

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend runs](https://gnss-analysis.swiftnav.com/jobs/id=974567)
+ [live-roof-650-townsend-dropouts-zero-baseline runs](https://gnss-analysis.swiftnav.com/jobs/id=974568)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-22-g062fb610&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-22-g062fb610&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-22-g062fb610&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-22-g062fb610&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019032221-22-g062fb610
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019032221-22-g062fb610)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019032221-22-g062fb610/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019032221-22-g062fb610

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019032221-23-g055f527f

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend runs](https://gnss-analysis.swiftnav.com/jobs/id=974897)
+ [live-roof-650-townsend-dropouts-zero-baseline runs](https://gnss-analysis.swiftnav.com/jobs/id=974898)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-23-g055f527f&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-23-g055f527f&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-23-g055f527f&build_type=buildroot_pr&display_type=table)
+ [live-roof-650-townsend-dropouts-zero-baseline](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-dropouts-zero-baseline&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019032221-23-g055f527f&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019032221-23-g055f527f
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019032221-23-g055f527f)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019032221-23-g055f527f/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019032221-23-g055f527f

File: package/sbp_fileio_daemon/overlay/etc/init.d/S81sbp_fileio_daemon_external [hunk 1]
```diff
@@ -14,8 +14,18 @@ file_perm=0644
...
+--pub-s ipc:///var/run/sockets/fileio/tcp_server1.pub \
+--sub-s ipc:///var/run/sockets/fileio/tcp_server1.sub \
+--pub-s ipc:///var/run/sockets/fileio/tcp_client0.pub \
+--sub-s ipc:///var/run/sockets/fileio/tcp_client0.sub \
+--pub-s ipc:///var/run/sockets/fileio/tcp_client1.pub \
+--sub-s ipc:///var/run/sockets/fileio/tcp_client1.sub \
+--pub-s ipc:///var/run/sockets/fileio/uart0.pub \
+--sub-s ipc:///var/run/sockets/fileio/uart0.sub \
+--pub-s ipc:///var/run/sockets/fileio/uart1.pub \
+--sub-s ipc:///var/run/sockets/fileio/uart1.sub \
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/overlay/etc/init.d/S81sbp_fileio_daemon_external]: Also added a comment about this in the S81sbp_fileio_daemon_external start-up script

**Pull request merged**